File: 0501-0600/0542-01 Matrix/0542-01 Matrix.py
import collections
import math
class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        m, n = len(matrix), len(matrix[0])
        dist = [[math.inf] * n for _ in range(m)]
        Q = collections.deque()
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    dist[i][j] = 0
                    Q.append((i, j))
        
        while Q:
            r, c = Q.popleft()
            for nr, nc in (r-1, c), (r+1, c), (r, c-1), (r, c+1):
                if 0 <= nr < m and 0 <= nc < n and dist[nr][nc] > dist[r][c] + 1:
                    dist[nr][nc] = dist[r][c] + 1
                    Q.append((nr, nc))
        return dist


class Solution2:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        m, n = len(matrix), len(matrix[0])
        dist = [[math.inf] * n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    dist[i][j] = 0
                else:
                    if i > 0:
                        dist[i][j] = min(dist[i][j], dist[i - 1][j] + 1)
                    if j > 0:
                        dist[i][j] = min(dist[i][j], dist[i][j - 1] + 1)
        
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if matrix[i][j] == 1:
                    if i < m - 1:
                        dist[i][j] = min(dist[i][j], dist[i + 1][j] + 1)
                    if j < n - 1:
                        dist[i][j] = min(dist[i][j], dist[i][j + 1] + 1)
        return dist

File: 0501-0600/0542-01 Matrix/0542-01 Matrix.cpp
// BFS
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {
        int m = matrix.size();        
        int n = matrix[0].size();
        queue<pair<int, int>> Q;
        vector<vector<int>> dist(m, vector<int>(n, INT_MAX));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] == 0) {
                    Q.emplace(i, j);
                    dist[i][j] = 0;
                }
            }
        }
        
        int dr[4] = {-1, 1, 0, 0};
        int dc[4] = {0, 0, -1, 1};
        while (!Q.empty()) {
            int row = Q.front().first;
            int col = Q.front().second;
            Q.pop();
            
            for (int k = 0; k < 4; ++k) {
                int nrow = row + dr[k];
                int ncol = col + dc[k];
                if (nrow >= 0 && nrow < m && ncol >= 0 && ncol < n && dist[nrow][ncol] > dist[row][col] + 1) {
                    dist[nrow][ncol] = dist[row][col] + 1;
                    Q.emplace(nrow, ncol);
                }
            }
        }
        
        return dist;
    }
};

// DP
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {
        int m = matrix.size();        
        int n = matrix[0].size();
        vector<vector<int>> dist(m, vector<int>(n, INT_MAX - 10000));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] != 0) {
                    if (j > 0) {
                        dist[i][j] = min(dist[i][j], 1 + dist[i][j - 1]);
                    } 
                    if (i > 0) {
                        dist[i][j] = min(dist[i][j], 1 + dist[i - 1][j]);
                    }
                }
                else {
                    dist[i][j] = 0;
                }
            }
        }
                
        for (int i = m - 1; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {
                if (matrix[i][j] != 0) {
                    if (j < n - 1) {
                        dist[i][j] = min(dist[i][j], 1 + dist[i][j + 1]);
                    } 
                    if (i < m - 1) {
                        dist[i][j] = min(dist[i][j], 1 + dist[i + 1][j]);
                    }
                }
            }
        }
        
        return dist;
    }
};
